### PR TITLE
ref(js): Make eventView query params compatible with react-router 6

### PR DESCRIPTION
**Pull request by evanpurkhiser**
The reason this has to change is due to react router 6 normalizing these in it's calls to `.push`. react router 3 does NOT do this normalization so it just passes through whatever it is called with.

To make things compatible we HAVE to do this translation. A lot of this can be removed once we migrate fully to react-router 6

### CODE REVIEW

File: static/app/utils/discover/eventView.tsx [hunk 1]
```diff
@@ -675,8 +700,8 @@ class EventView {
       field: this.getFields(),
       widths: this.getWidths(),
       sort: encodeSorts(this.sorts),
-      environment: this.environment,
-      project: this.project,
+      environment: [...this.environment],
+      project: [...this.project],
```
Comment (evanpurkhiser) [inline static/app/utils/discover/eventView.tsx:704]: Needs to be cloned since it's mutated by `stringifyQueryParams`

**Review by gggritso — COMMENTED** [review]
Is the idea that RR6's typing doesn't let it accept single-length arrays, or numbers? Does it break in some specific way? I'm not following from the PR description why this is needed

File: static/app/utils/discover/eventView.tsx [hunk 2]
```diff
@@ -177,6 +177,31 @@ export const encodeSort = (sort: Sort): string => {
...
+        query[field] = query[field][0];
+      }
+      if (query[field].length === 0) {
+        query[field] = undefined;
+      }
+    } else {
+      query[field] = query[field]?.toString();
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (evanpurkhiser) [reply static/app/utils/discover/eventView.tsx:203]: I don't know if it's THAT scary

**Comment by evanpurkhiser** [comment]
>Is the idea that RR6's typing doesn't let it accept single-length arrays, or numbers? Does it break in some specific way? I'm not following from the PR description why this is needed

No rr6 accepts this stuff just fine. It's just that react router 3 does not normalize these values when it calls `history.push` while react router 6 **does** normalize these things. So to make things work between 3 and 6 we have to expect and pass values that look like what react router 6 would normalize to for the call to `.push`

File: static/app/utils/discover/eventView.tsx [hunk 3]
```diff
@@ -177,6 +177,31 @@ export const encodeSort = (sort: Sort): string => {
...
+      query[field] = query[field].map((v: string | number) => v.toString());
+
+      if (query[field].length === 1) {
+        query[field] = query[field][0];
+      }
+      if (query[field].length === 0) {
+        query[field] = undefined;
+      }
+    } else {
+      query[field] = query[field]?.toString();
```
[the remark this answers is not in the document]
Comment (evanpurkhiser) [reply static/app/utils/discover/eventView.tsx:200]: should be fine

**Review by shruthilayaj — APPROVED** [review]
This looks good to me - I also spot checked some of the places where we use these generated target urls, they work fine
There is this one [spot](https://github.com/getsentry/sentry/pull/76461/files#diff-2d6b37954a3d6e74111ad71f6ebbf2d4462b52d5b1a102edea0205e399a8991fR1223-R1225) in eventView where we modify the params after they've been stringified - does this also require a `.toString()`?

**Comment by evanpurkhiser** [comment]
> There is this one [spot](https://github.com/getsentry/sentry/pull/76461/files#diff-2d6b37954a3d6e74111ad71f6ebbf2d4462b52d5b1a102edea0205e399a8991fR1223-R1225) in eventView where we modify the params after they've been stringified - does this also require a .toString()?

Nah, if there's no test for it it doesn't matter. Everything still _worked_ between 3 and 6, but to test it was slightly different.

**Pull request merged**